### PR TITLE
more cleanup in preparation for model translation

### PR DESCRIPTION
**Pull request by Goober5000**
1) Clean up stepped rotation, fixing some logic, adding safeguards for invalid values, and allowing rotations to go in the opposite direction
2) Add more specific SEXP error checking for AWACS and rotating subsystems
3) Assorted refactoring and polishing

### CODE REVIEW

File: code/parse/sexp.cpp [hunk 1]
```diff
@@ -20250,28 +20250,40 @@ void sexp_set_subsys_rotation_lock_free(int node, bool locked)
...
-				rotate->subsys_snd_flags.set(Ship::Subsys_Sound_Flags::Rotate);
+				obj_snd_assign(ship_entry->shipp->objnum, subsys->system_info->rotation_snd, &subsys->system_info->pnt, OS_SUBSYS_ROTATION, subsys);
+				subsys->subsys_snd_flags.set(Ship::Subsys_Sound_Flags::Rotate);
 			}
+
+			// and restore the timestamp from the delta
+			auto &stamp = subsys->submodel_instance_1->stepped_rotation_started;
+			if (stamp.isValid())
+				stamp = timestamp_delta(_timestamp(), -stamp.value());
 		}
```
Comment (BMagnu) [inline code/parse/sexp.cpp:20287]: This whole chunk will cause a whole load of trouble if someone locks or unlocks a subsystem that was already locked or unlocked respectively.
You'll need to at least safeguard it by checking if the current state is actually different before modifying the timestamp

**Review by BMagnu — CHANGES_REQUESTED** [review]
Mostly good cleanup, but a few minor things

File: code/model/modelread.cpp [hunk 2]
```diff
@@ -788,48 +789,72 @@ static void set_subsystem_info(int model_num, model_subsystem *subsystemp, char
...
 			float fraction = subsystemp->stepped_rotation->fraction;
 
-			subsystemp->stepped_rotation->max_turn_accel = PI2 / (fraction*(1.0f - fraction) * num_steps * t_trans*t_trans);
-			subsystemp->stepped_rotation->max_turn_rate =  PI2 / ((1.0f - fraction) * num_steps *t_trans);
+			// reverse the direction if we start out with reverse velocity
+			if (turn_rate < 0.0f) {
+				subsystemp->stepped_rotation->backwards = true;
+			}
+
+			subsystemp->stepped_rotation->max_turn_accel = (fraction == 0.0f) ? 0.0f : step_distance / (fraction * (1.0f - fraction) * t_trans * t_trans);
```
[the remark this answers is not in the document]
Comment (Goober5000) [reply code/model/modelread.cpp:856]: Oh, good point

File: code/model/modelread.cpp [hunk 3]
```diff
@@ -3791,24 +3810,14 @@ void submodel_stepped_rotate(model_subsystem *psub, submodel_instance *smi)
...
 
-	if (cur_step >= psub->stepped_rotation->num_steps) {
-		// I don;t know why, but removing this line makes it all good.
-		// sii->turn_step_zero_timestamp += int(1000.0f * (psub->stepped_rotation->num_steps * step_time) + 0.5f);
-
-		// reset cur_step (use mod to handle physics/ai pause)
-		cur_step = cur_step % psub->stepped_rotation->num_steps;
-	}
+	// get step we are on
+	int cur_step = static_cast<int>(floor(elapsed_time / step_time));
```
[the remark this answers is not in the document]
Comment (Goober5000) [reply code/model/modelread.cpp:3817]: Ok.

File: code/parse/sexp.cpp (hunk 1 above)
Comment (Goober5000) [reply code/parse/sexp.cpp:20287]: Ok.

**Comment by Goober5000** [comment]
Ok, changes addressed.

File: code/math/floating.h [hunk 4]
```diff
@@ -96,23 +96,22 @@ float fl_isqrt( float x )
 } 
 */
 
+// sees if a floating point number is within the minimum tolerance of zero
+inline bool fl_near_zero(float a)
+{
+	return a < std::numeric_limits<float>::epsilon()
+		|| a > -std::numeric_limits<float>::epsilon();
```
Comment (BMagnu) [inline code/math/floating.h:103]: Wouldn't this need to be an &&?

**Review by BMagnu — CHANGES_REQUESTED** [review]

File: code/math/floating.h (hunk 4 above)
Comment (Goober5000) [reply code/math/floating.h]: Whoops.  Yes.  Pushed again.

**Review by BMagnu — APPROVED** [review]
Looks good now

**Pull request merged**